### DISCUSSION TITLE
Extension Interacting with a Localhost Server does not work in a VS Code cloud environment

### DISCUSSION THREAD

**Discussion by caleb-at-pieces** [microsoft/vscode-discussions#873] [Extension Development QnA] [answered]
Hello,

I am on the development team at https://pieces.app . We have a plethora of extensions that all work by interacting with a local service, called Pieces OS, that runs on the user's machine. This all works very well until the user is utilizing some sort of cloud shell environment in VS Code, i.e a dev container, a GitHub Codespace, or an ssh tunnel, as the extension is trying to reach to localhost in order to interact with our Pieces OS api's, and Pieces OS is running on the user's machine, not the development environment.

Question:

Is there some way, either for our users or for us as the extension developers, to enable VS Code to interact with a server that is running on the user's machine from their VS Code cloud shell.

**Comment by tamuratak** [comment, marked as answer]
Reverse tunnel, accessing from a remote machine to a local machine, is not supported. The feature request is tracked at microsoft/vscode/issues/100222.

Possible workarounds:

- By setting `"extensionKind": ["ui"]`, your extension always runs on the user's local machine. See https://code.visualstudio.com/api/advanced-topics/extension-host#preferred-extension-location. This works only for VS Code Desktop.
- Use Local Port Forwarding, accessing from a remote machine to a local machine via `*.tunnels.api.visualstudio.com`. See https://code.visualstudio.com/docs/editor/port-forwarding
